### PR TITLE
feat(configuration): Add support for spring cloud config with S3

### PR DESCRIPTION
**Pull request by titirigaiulian**
Due to corporate restrictions we cannot add git as a backend for spring-cloud-config. In the latest releases of spring cloud, S3 was added as supported backend. It would be great to use S3 for storing both clouddriver account definitions and kubeconfigs. 

This PR starts from a previous one [here](https://github.com/spinnaker/kork/pull/455) and reverted [here](https://github.com/spinnaker/kork/pull/517) due to the following [issue](https://github.com/spinnaker/spinnaker/issues/5449)

The issue identified is that when using `spring-cloud-starter-aws` it includes `spring-cloud-aws-autoconfigure` with lots of Autoconfiguration [classes](https://github.com/spring-cloud/spring-cloud-aws/blob/master/spring-cloud-aws-autoconfigure/src/main/resources/META-INF/spring.factories) causing failures at startup for non aws environments.

Approach: Use only `spring-cloud-aws-contex` and configure the `ContextResourceLoaderConfiguration` for S3

Tested on spinnaker installed in an EKS cluster with latest clouddriver and the following config:
`spinnakerconfig.yml`
```
spring:
  profiles:
    include: awss3
  cloud:
    config:
      server:
        awss3:
          region: us-east-1
          bucket: <my-bucket>
```
The s3 bucket `my-bucket` contains a `clouddriver-local.yml` with K8s cluster definitions and reference to kubeconfig files `kubeconfigFile: configserver:my_kubeconfig.yml` and under `my-bucket/clouddriver/` the kubeconfig files: `my_kubeconfig.yml`

ISSUE: https://github.com/spinnaker/spinnaker/issues/5532

### CODE REVIEW

File: kork-config/src/main/java/com/netflix/spinnaker/kork/configserver/autoconfig/SpringCloudAwsConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+@Profile("awss3")
+@Configuration
+class SpringCloudAwsConfiguration {
+
+  @Bean
+  public static AmazonS3Client amazonS3(AwsS3EnvironmentProperties s3EnvironmentProperties) {
+    return (AmazonS3Client)
+        AmazonS3ClientBuilder.standard()
+            .withRegion(s3EnvironmentProperties.getRegion())
+            .withCredentials(new DefaultAWSCredentialsProviderChain())
```
Comment (costimuraru) [inline kork-config/src/main/java/com/netflix/spinnaker/kork/configserver/autoconfig/SpringCloudAwsConfiguration.java]: Is this needed? I believe if we skip this line, the default credential is used either way.

**Comment by scottfrederick** [comment]
> Approach: Use only `spring-cloud-aws-context` and configure the `ContextResourceLoaderConfiguration` for S3

This is a good approach @titirigaiulian, and should simplify things quite a bit. Thanks for working through the details to get this working.

File: kork-config/src/main/java/com/netflix/spinnaker/kork/configserver/autoconfig/SpringCloudAwsConfiguration.java (hunk 1 above)
Comment (titirigaiulian) [reply kork-config/src/main/java/com/netflix/spinnaker/kork/configserver/autoconfig/SpringCloudAwsConfiguration.java]: done

File: spinnaker-dependencies/spinnaker-dependencies.gradle [hunk 2]
```diff
@@ -58,6 +58,7 @@ dependencies {
   api(platform("org.springframework.boot:spring-boot-dependencies:${versions.springBoot}"))
   api(platform("com.amazonaws:aws-java-sdk-bom:${versions.aws}"))
   api(platform("org.springframework.cloud:spring-cloud-dependencies:${versions.springCloud}"))
+  api(platform("com.amazonaws:aws-java-sdk-s3:${versions.aws}"))
```
[the remark this answers is not in the document]
Comment (titirigaiulian) [reply spinnaker-dependencies/spinnaker-dependencies.gradle]: done

File: spinnaker-dependencies/spinnaker-dependencies.gradle [hunk 3]
```diff
@@ -123,7 +124,7 @@ dependencies {
     api("commons-io:commons-io:2.5")
     api("de.danielbechler:java-object-diff:0.95")
     api("de.huxhorn.sulky:de.huxhorn.sulky.ulid:8.2.0")
-    api("dev.minutest:minutest:1.11.0")
+    api("dev.minutest:minutest:1.10.0")
```
[the remark this answers is not in the document]
Comment (titirigaiulian) [reply spinnaker-dependencies/spinnaker-dependencies.gradle]: Rebase leftover. Done

**Comment by titirigaiulian** [comment]
If it helps, I published an image on dockerhub for testing with these changes on top of the latest master in clouddriver: `iuliantitiriga/clouddriver:ext_conf_S3`

**Comment by titirigaiulian** [comment]
@robzienert @cfieber can you please take a look and merge it?

File: kork-config/kork-config.gradle [hunk 4]
```diff
@@ -11,6 +11,8 @@ dependencies {
 
   api "org.springframework.cloud:spring-cloud-context"
   api "org.springframework.cloud:spring-cloud-config-server"
+  api "org.springframework.cloud:spring-cloud-aws-context"
```
Comment (cfieber) [inline kork-config/kork-config.gradle:14]: I believe these should be `implementation` not `api`

**Comment by cfieber** [comment]
I don't really feel comfortable weighing in on this - this isn't something we use so I'd defer to someone in the community that is running with config server enabled

**Comment by titirigaiulian** [comment]
thanks @cfieber. Do you know where can I follow up. What SIG is suitable?

**Comment by robzienert** [comment]
Platform or Ops.

**Comment by dragon3** [comment]
I was trying to use S3 for external account configuration backend as following this document:
https://spinnaker.io/setup/configuration/#using-an-s3-backend

But it didn't work and I found the issue https://github.com/spinnaker/spinnaker/issues/5532 and this pull request.

Do you have any plan to merge and release this change in the near future?

( If no plan, the current document is very confusing since it's not actually available 😢 )

**Comment by robzienert** [comment]
I'll merge this if the PR is updated. Will close in a week if not.

**Comment by titirigaiulian** [comment]
Rebased and tested it with latest `clouddriver` on spinnaker v.1.22.1 on an EKS cluster

**Review by robzienert — APPROVED** [review]

**Pull request merged**